packages/fxa-auth-server/fxa-oauth-server/lib/jwt_sub.js

```diff
@@ -0,0 +1,44 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const hex = require('buf').to.hex;
+const hkdf = require('../../lib/crypto/hkdf');
+const config = require('./config');
+
+const PPID_ENABLED = config.get('ppid.enabled');
+const PPID_CLIENT_IDS = new Set(config.get('ppid.enabledClientIds'));
+const PPID_ROTATING_CLIENT_IDS = new Set(config.get('ppid.rotatingClientIds'));
+const PPID_ROTATION_PERIOD_MS = config.get('ppid.rotationPeriodMS');
+const PPID_SALT = config.get('ppid.salt');
+const PPID_INFO = 'oidc ppid sub';
+
+const SUB_LENGTH_BYTES = config.get('unique.id');
+
+module.exports = async function generateSub(
+  userIdBuf,
+  clientIdBuf,
+  clientSeed = 0
+) {
+  const userIdHex = hex(userIdBuf);
+  const clientIdHex = hex(clientIdBuf);
+
+  // TODO - constrain clientSeed.
+
+  if (PPID_ENABLED && PPID_CLIENT_IDS.has(clientIdHex)) {
```

Comment:
@martinthomson, could I get your feedback on whether is reasonable/matches your expectations?

Reply:
Is the idea here that you would configure PPID as a global setting and that it could be requested on a request-by-request basis?

Because I would have thought that if PPID was enabled, you would want an error if there was no client-provided ID.

Reply:
> Is the idea here that you would configure PPID as a global setting and that it could be requested on a request-by-request basis?

The idea is that PPIDs are enabled on both a global and per client basis. Per-client only matters if enabled on a global basis. We do this sort of gating often in FxA to quickly disable a feature globally if it is found to cause some sort of problem.

> Because I would have thought that if PPID was enabled, you would want an error if there was no client-provided ID.

The only way to get to this function is via the /token endpoint, which requires a [client_id to be passed](https://github.com/mozilla/fxa/pull/1679/files#diff-19a8f6d558692592349f678e6cc65b38R64) in either the Authorization header or POST request body.

Reply:
Note to self, update the /authorization endpoint w/ ppid_seed and ensure the client_id is required when generating access tokens.